ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java

```diff
@@ -135,6 +152,177 @@ public static void updateCrosslinks(BeaconState state) throws Exception {
     }
   }
 
+  /**
+   * Rewards and penalties applied with respect to justification and finalization. Spec:
+   * https://github.com/ethereum/eth2.0-specs/blob/v0.1/specs/core/0_beacon-chain.md#justification-and-finalization
+   *
+   * @param state
+   */
+  public static void justificationAndFinalization(BeaconState state) throws Exception {
+
+    final UnsignedLong FOUR = UnsignedLong.valueOf(4L);
+    UnsignedLong epochs_since_finality =
+        BeaconStateUtil.get_next_epoch(state).minus(state.getFinalized_epoch());
+    UnsignedLong previous_total_balance = previous_total_balance(state);
+    List<UnsignedLong> balances = state.getValidator_balances();
+
+    // Case 1: epochs_since_finality <= 4:
+    if (epochs_since_finality.compareTo(FOUR) <= 0) {
+      // Expected FFG source
+      UnsignedLong previous_balance =
+          AttestationUtil.get_previous_epoch_justified_attesting_balance(state);
+      List<Integer> previous_indices =
+          AttestationUtil.get_previous_epoch_justified_attester_indices(state);
+      case_one_penalties_and_rewards(
+          state, balances, previous_total_balance, previous_balance, previous_indices);
+
+      // Expected FFG target
+      previous_balance = AttestationUtil.get_previous_epoch_boundary_attesting_balance(state);
+      previous_indices = AttestationUtil.get_previous_epoch_boundary_attester_indices(state);
```

Comment:
I think the repeated use of same variable for possibly different lists is a bit confusing for the reader, altough might not be a concern right now.

Reply:
Possibly, but I did it so that we did keep 4x index lists in memory for no reason